Sprint 7 solution in file manager

SlavaTurovski opened the pull request.

SlavaTurovski committed:
создал новый класс Node

SlavaTurovski committed:
изменил механику хранения истории

SlavaTurovski committed:
изменил механику хранения истории

SlavaTurovski committed:
изменил механику хранения истории

SlavaTurovski committed:
Добавил новые юнит тесты и обновил старые

SlavaTurovski committed:
Добавил новые юнит тесты и обновил старые

SlavaTurovski committed:
Добавил новые юнит тесты и обновил старые

SlavaTurovski committed:
Правки по коду

SlavaTurovski committed:
Правки по коду

SlavaTurovski committed:
Правки по коду

SlavaTurovski committed:
Правки по коду

SlavaTurovski committed:
Правки по коду

SlavaTurovski committed:
Основная доработка проекта по ТЗ-7

SlavaTurovski committed:
Правки code style

tomattman commented on src/.DS_Store:
что это за файл?

tomattman commented on src/Main.java:
```diff
@@ -1,19 +1,20 @@
+/*import interfaces.TaskManager;
```
закомментированный код не допускается

tomattman commented on src/Main.java:
```diff
@@ -1,19 +1,20 @@
 public class Main {
 
-    static InMemoryTaskManager taskManager = (InMemoryTaskManager) Managers.getDefault();
+    /*static InMemoryTaskManager taskManager = (InMemoryTaskManager) Managers.getDefault();*/
```
комментарий к предыдущей версии зачем приводить к InMemoryTaskManager ведь мы можем использовать по интерфейсу, это одно из их преимуществ и реализация инкапсуляции, нам не важно что спрятано под интерфейсом, нам важно что то, что спрятано под ним умеет выполнять его действия

tomattman commented on src/manager/CSVTaskFormat.java:
```diff
@@ -0,0 +1,51 @@
+
+public class CSVTaskFormat {
+
+    public static String toString(Task task) {
```
лучше сделать через перегрузку
сделать два метода с одинаковыми именами, один принимает как параметр Task второй принимает сабтаск
JVM умная, поймет что когда мы вызываем метод и в параметрах эпик или таск, то выполнит метод в котором таск (как самый подходящий) а если в параметрах будет сабтаск, ты выполнит метод в котором в параметрах сабтаск

tomattman commented on src/manager/CSVTaskFormat.java:
```diff
@@ -0,0 +1,51 @@
+public class CSVTaskFormat {
+
+    public static String toString(Task task) {
+        StringBuilder result = new StringBuilder();
```
маленькая мелочь, как можно реализовать
StringBuilder result = new StringBuilder()
.append(task.getId()).append(",")
.append(task.getTaskType()).append(",")
и так далее

tomattman replied:
или даже так
return new StringBuilder()
.append(task.getId()).append(",")
.append(task.getTaskType()).append(",")
//и так далее
.toString()

tomattman commented on src/manager/FileBackedTaskManager.java:
```diff
@@ -0,0 +1,212 @@
+
+    private final File file;
+
+    public FileBackedTaskManager(File file) {
```
еще вариант улучшение
Если сделать конструктор приватным, то его можно будет вызвать только из методов класса, в конкретно этом случае, только из статического метода, а следовательно, создать экземпляр этого класса получится только с попыткой чтения из файла, а следовательно меньше шанс потерять данные при создании экземпляра класса

tomattman commented on src/manager/FileBackedTaskManager.java:
```diff
@@ -0,0 +1,212 @@
+            while ((line = reader.readLine()) != null) {
+                Task task = fromString(line);
+
+                if (task instanceof Epic) {
```
instanceof ооп подход, его не следует использовать
у тебя хранится тип таска в самом классе, можно отталкиваться от него, ну или снова через перегрузку

tomattman commented on src/manager/FileBackedTaskManager.java:
```diff
@@ -0,0 +1,212 @@
+                if (task instanceof Epic) {
+                    manager.addEpic((Epic) task);
+                } else if (task instanceof Subtask) {
+                    manager.addSubtask((Subtask) task);
```
не стоит использовать метод на добавление , потому что он заменят идентификатор и у тебя теряется связь между сабтасками и эпиками

tomattman commented on src/manager/FileBackedTaskManager.java:
```diff
@@ -0,0 +1,212 @@
+
+    @Override
+    public Task getTaskById(int id) {
+        return super.getTaskById(id);
```
методы которые не меняют поведение класса родителя, можно не переопределять

tomattman commented on src/manager/InMemoryHistoryManager.java:
```diff
@@ -4,27 +4,90 @@
 
-    private static final int LIMIT_TASKS = 10;
-    private final List<Task> historyTasks = new ArrayList<>();
+    private final HashMap<Integer, Node> history = new HashMap<>();
```
ты не смерджил предыдущий пулл реквест, поэтому у тебя в текущем видны изменения из предыдущего,
перед следующей сдачей не забудь его смерджить, чтобы в этом пулл реквесте не были видны изменения из него

tomattman commented on src/manager/InMemoryTaskManager.java:
```diff
@@ -28,6 +25,10 @@ private int generateId() {
         return id++;
     }
 
+    protected void setActualId(int actualId) {
```
зачем нужен этот метод

tomattman commented on src/manager/InMemoryTaskManager.java:
```diff
@@ -244,8 +266,20 @@ public List<Subtask> getSubtasksByEpicId(int id) {
+        return (ArrayList<Task>) historyManager.getHistory();
+    }
+
+    protected TaskType getTaskTypeById(int id) {
```
и зачем нужен этот метод

tomattman commented on src/manager/Managers.java:
```diff
@@ -3,14 +3,16 @@
 
     public static TaskManager getDefault() {
-        return new InMemoryTaskManager();
+        return FileBackedTaskManager.loadFromFile(new File("tasks.csv"));
```
а нужно ли, чтобы по умолчанию теперь возвращался файловый таск менеджер?

tomattman reviewed (changes requested):
Привет
1) тебе надо поработать с файлом .gitignore много файлов которые есть в пулл реквесте не должны быть в нем, например папка out генерируется при билде проекта, и такие папки добавлять не стоит 
2) стоит смерджить предыдущий пулл реквест, чтобы его изменения не отображались в этом
3) остальные комменты в коде, но в целом работа хорошая

SlavaTurovski committed:
Правки замечаний после ревью

SlavaTurovski committed:
Delete out/production/java-kanban directory

SlavaTurovski committed:
Delete .DS_Store

SlavaTurovski committed:
Delete src/.DS_Store

SlavaTurovski committed:
Правки замечаний после ревью

SlavaTurovski committed:
Merge branch 'sprint_7-solution-in-file-manager' of https://github.com/SlavaTurovski/java-kanban into sprint_7-solution-in-file-manager

SlavaTurovski committed:
Delete .DS_Store

SlavaTurovski committed:
Правки после ревью

SlavaTurovski committed:
Merge branch 'sprint_7-solution-in-file-manager' of https://github.com/SlavaTurovski/java-kanban into sprint_7-solution-in-file-manager

tomattman commented on src/manager/CSVTaskFormat.java:
```diff
@@ -0,0 +1,51 @@
+    }
+
+    public static String toString(Subtask subtask) {
+        return toString((Task) subtask) + "," + subtask.getEpicId();
```
прекрасное решение

tomattman commented on src/manager/FileBackedTaskManager.java:
```diff
@@ -0,0 +1,194 @@
+
+    public static FileBackedTaskManager loadFromFile(File file) {
+
+            FileBackedTaskManager manager = new FileBackedTaskManager(file);
```
осталось две проблемы
1) счетчик идентификаторов не изменяется, следовательно после считывания из файла, есть шанс что пи добавлении перетрется старое значение
воспроизвести следующим образом
создаешь файловый менеджер, добавляешь в него таск, создаешь новый файловый менеджер к тому же файлу, смотришь что в нем один таск, добавляешь новый таск, проверяешь что в находится в мапе, должно быть два таска
2) после считывания из файла, у эпика будет пустой список сабтасок

tomattman reviewed (changes requested):
Привет, гораздо лучше, но осталось парочка проблем при считывании

SlavaTurovski committed:
Правки после ревью

- внес поправки по механике загрузки
- фиксил счетчик
- фиксил отсутствие id сабтасок в эпиках

tomattman reviewed (approved):
Принято
